bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java

```diff
@@ -84,24 +83,20 @@
 
     public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_BRIDGE);
 
-    private static final String DEVICE_TYPE = "EclipseSmartHome";
+    private static final long BYPASS_MIN_DURATION_BEFORE_CMD = 1500L;
 
-    private static enum StatusType {
-        ADDED,
-        REMOVED,
-        GONE,
-        CHANGED
-    }
+    private static final String DEVICE_TYPE = "EclipseSmartHome";
 
     private final Logger logger = LoggerFactory.getLogger(HueBridgeHandler.class);
 
     private final Map<String, FullLight> lastLightStates = new ConcurrentHashMap<>();
     private final Map<String, FullSensor> lastSensorStates = new ConcurrentHashMap<>();
     private final Map<String, FullGroup> lastGroupStates = new ConcurrentHashMap<>();
 
-    private final List<LightStatusListener> lightStatusListeners = new CopyOnWriteArrayList<>();
-    private final List<SensorStatusListener> sensorStatusListeners = new CopyOnWriteArrayList<>();
-    private final List<GroupStatusListener> groupStatusListeners = new CopyOnWriteArrayList<>();
+    private @Nullable HueLightDiscoveryService discoveryService;
+    private final Map<String, LightStatusListener> lightStatusListeners = new ConcurrentHashMap<>();
```

Comment:
Why not use the `ThingUID` instead of device is (`String`) as key? This way we can think about a even greater refactoring and combine all different listener interfaces into one single `StatusListener` interface. Is that possible?

Another idea which came to my mind was to improve registration of the listeners by using [BridgeHandler#childHandlerInitialized](https://github.com/openhab/openhab-core/blob/ac1f4080e3896d0a9b44cef3e5ca74efa58273ed/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/BridgeHandler.java#L38) method. This way the framework takes care of whole initialization process of bridge and things in the correct order and we can completely remove the logic from the Hue binding.

Reply:
I already thought about a Generic StatusListener interface. I can give this a try this evening. And one thing I although thought about is to move update(Light|Sensor|Group)State methods to the belonging things and reverse the logic so they get the hueBridge from inside. Because in my opinion this is mainly logic from inside. 

> Why not use the `ThingUID` instead of device is (String) as key? 

What is not clear to me is how the polling jobs can determine the right listener then? Do you have an example?

Reply:
I thought about this again. And maybe we should do this on a new PR for two reasons:
First this PR is mainly about fixing the polling issue after command, and I think this maybe intransparent to hide this huge refactoring behind.
Second this huge refactoring will surely take some time with all needed reviews and discussions. That in my opinion we should bring this fix to next release first, before this starting the process of refactoring. 

What do you think about this @cweitkamp?

Reply:
> What is not clear to me is how the polling jobs can determine the right listener then?

You probably have to change the related methods in the listener interfaces too. They can return the `ThingUID` as well.

> What do you think about this @cweitkamp?

Of course. Feel free to create as much PRs as you like.

Reply:
> You probably have to change the related methods in the listener interfaces too. They can return the `ThingUID` as well.

The registration of the listener is not my problem on this point. That's easy, like you suggested for example the LightListener can return the ThingUID instead of the lightID. But how can the light polling job inside the bridgeHandler find the correct LightThingHandler in the map. 

I don't see a problem when users use the ThingUID suggested by the system in there configuration like this
`0220 1 "LED Livingroom" @ "Controls" [ lightId="1" ]` 

In this case the polling job can automatically  generate the ThingUID with the information it got from hardware and select the right listener from Map to update the state.

But if users use something like this in there configuration:
`0220 livingroomled "LED Livingroom" @ "Controls" [ lightId="1" ]` 

How can pollingJob select in this case the correct listener from Map without iterate over all listeners again, because iteration is the thing I want to avoid with the map?

Reply:
> Of course. Feel free to create as much PRs as you like.

I know but can we finish this one first, so I can base further refactoring on the result of this PR.

Reply:
Thanks for the detailed description. I finally got the point. We of course cannot bet on the users way to name his ThingUID.